src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/Types/DynamicContentFieldsIndexAliasProvider.cs

```diff
@@ -0,0 +1,120 @@
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using GraphQL;
+using Microsoft.Extensions.Options;
+using OrchardCore.ContentManagement.GraphQL.Options;
+using OrchardCore.ContentManagement.Metadata;
+using OrchardCore.ContentTypes.Events;
+using OrchardCore.Environment.Shell;
+
+namespace OrchardCore.ContentManagement.GraphQL.Queries.Types;
+
+public class DynamicContentFieldsIndexAliasProvider : IIndexAliasProvider, IContentDefinitionEventHandler
+{
+    private static readonly ConcurrentDictionary<string, List<IndexAlias>> _aliases = new ConcurrentDictionary<string, List<IndexAlias>>();
+
+    private readonly IContentDefinitionManager _contentDefinitionManager;
+    private readonly IEnumerable<IContentFieldProvider> _contentFieldProviders;
+    private readonly ShellSettings _shellSettings;
+    private readonly GraphQLContentOptions _contentOptions;
+
+    public DynamicContentFieldsIndexAliasProvider(IContentDefinitionManager contentDefinitionManager,
+        IEnumerable<IContentFieldProvider> contentFieldProviders,
+        IOptions<GraphQLContentOptions> contentOptionsAccessor,
+        ShellSettings shellSettings)
+    {
+        _contentDefinitionManager = contentDefinitionManager;
+        _contentFieldProviders = contentFieldProviders;
+        _contentOptions = contentOptionsAccessor.Value;
+        _shellSettings = shellSettings;
+    }
+
+    public IEnumerable<IndexAlias> GetAliases()
+    {
+        var tenantAliases = _aliases.GetOrAdd(_shellSettings.Name, _ => []);
+
+        if (tenantAliases.Count != 0)
```

Comment:
This is not thread-safe. Clearing and populating the tenantAliases could be called concurrently.

Reply:
@gvkries Done

Reply:
I think this is still not fully correct. I would suggest to move populating the aliases into the callback of `GetOrAdd()`, don't test for an empty collection and return the result right away. When clearing, just remove the entry and let the GC handle the rest (do not clear the inner list). That should be more simple to understand and is thread-safe, because the concurrent dictionary handles it completely.

Reply:
You are right when testing multiple scenarios, should be fixed now

Reply:
Agree with @gvkries who mentioned that the filler logic should be called in GetOrAdd, e.g. move all the logic underneath this comment inside the callback method

return _aliases.GetOrAdd(_shellSettings.Name,_=>{
// the logic below that method
})

As a side note, this type of caching using tenant names as keys is not very common in OC source code.

It is recommended to refer to `ContentDefinitionManager` to use `IMemoryCache` , which has been designed to be tenant-isolated in OC.